src/main/java/net/minestom/server/entity/EntityQuery.java

```diff
@@ -0,0 +1,169 @@
+package net.minestom.server.entity;
+
+import net.minestom.server.coordinate.Point;
+import net.minestom.server.coordinate.Vec;
+import net.minestom.server.entity.EntityQueryImpl.Index;
+import net.minestom.server.utils.Range;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.function.BiPredicate;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * Descriptive how entities should be queried.
+ * Offer potential indexing/spatial partitioning advantages over lazy looping.
+ */
+public sealed interface EntityQuery permits EntityQueryImpl {
+    Property<Boolean> PLAYER = new Index<>();
+    Property<Integer> ID = new Index<>();
+    Property<UUID> UUID = new Index<>();
+    Property<String> NAME = new Index<>();
+    Property<Point> COORD = new Index<>();
+    Property<Double> POS_X = new Index<>();
+    Property<Double> POS_Y = new Index<>();
+    Property<Double> POS_Z = new Index<>();
+    Property<Float> YAW = new Index<>();
+    Property<Float> PITCH = new Index<>();
+    Property<Integer> CHUNK_X = new Index<>();
+    Property<Integer> CHUNK_Z = new Index<>();
+    Property<Double> DISTANCE = new Index<>();
+    Property<EntityType> TYPE = new Index<>();
+    Property<GameMode> GAME_MODE = new Index<>();
+    Property<Integer> EXPERIENCE = new Index<>();
```

Comment:
Users own properties would be helpful here

Reply:
Can leave it up to users to update probably (agreed on the safety issue but it's just user error at that point)